### PR TITLE
`circuit-types`, `circuit-macros`: Refactor implementation over Stark curve

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR refactors the implementations of the `circuit-macros` and `circuit-types` crates to operate over the Stark curve. This largely involves changing types, but the large changes include:
- Changing the traits in `circuit-types`; specifically allowing the underlying `async` methods to color our interfaces. This exposes an async flow over the fabric executor that allows the caller to await multiple operations at once that were previously individually blocking.
- Introduce a mock MPC testing setup that replaces a network connection with a `tokio::io::DuplexStream`.
- Remove the `<N: MpcNetwork, S: SharedValueSource<Scalar>>` generics from all interfaces, these are removed in the underlying `mpc-stark` library. This greatly simplifies generic code.

### Testing
- Crate unit tests pass for all modified crates; CI will be red until other crates are refactored.

### CODE REVIEW

File: circuit-types/src/macro_tests.rs [hunk 1]
```diff
@@ -0,0 +1,276 @@
+//! Defines tests for macros in the `circuit_macros` crate. We do this so that we may define the
+//! bulk of the traits, data structures, etc outside of the `circuit-macros` crate; as a proc-macro
+//! crate cannot export non proc-macro items
+
+#[allow(clippy::missing_docs_in_private_items)]
```
Comment (joeykraut) [inline circuit-types/src/macro_tests.rs:5]: These are mostly moved from the `circuits` crate. I did introduce a new MPC mocking setup here that is used in two of the tests though

File: test-helpers/src/mpc_network.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+/// The size of MPC network to allocate by default
+pub const DEFAULT_MPC_SIZE_HINT: usize = 1_000_000;
+
+// -----------
+// | Helpers |
+// -----------
+
+/// Sets up a basic MPC fabric between two parties using the QUIC transport and
+/// the beaver triplet source defined above
+pub fn setup_mpc_fabric(party_id: u64, local_port: u64, peer_port: u64, docker: bool) -> MpcFabric {
```
Comment (joeykraut) [inline test-helpers/src/mpc_network.rs:28]: this method was copied from the analogous method in `integration-helpers` crate, which was deleted in favor of this crate

File: circuit-macros/src/circuit_type/mpc_types.rs [hunk 3]
```diff
@@ -134,24 +121,37 @@ fn build_mpc_type(
...
-        ident_with_generics(new_ident(MPC_TYPE_SERIALIZED_IDENT), build_mpc_generics());
+    let authenticated_scalar_type = path_from_ident(new_ident(MPC_TYPE_SERIALIZED_IDENT));
     let from_auth_scalars_method = build_deserialize_method(
         new_ident(FROM_AUTHENTICATED_SCALARS_METHOD_NAME),
         authenticated_scalar_type.clone(),
         mpc_type_trait_name.clone(),
         mpc_type,
     );
 
+    // Build a `fabric` method
```
Comment (akirillo) [inline circuit-macros/src/circuit_type/mpc_types.rs:140]: hmm, relying on the ordering of fields feels a bit brittle, but idrk how you would more flexibly search for the first field that has a `fabric` method / implements some trait...

**Review by akirillo — APPROVED** [review]
overall lgtm, have some questions b/c i lack context

File: circuit-types/src/fixed_point.rs [hunk 4]
```diff
@@ -401,49 +405,16 @@ impl<L: LinearCombinationLike> Sub<Variable> for FixedPointVar<L> {
     }
 }
 
-impl<N: MpcNetwork + Send, S: SharedValueSource<Scalar>> AuthenticatedFixedPoint<N, S> {
-    /// Create an authenticated fixed-point variable from a given scalar integer
-    pub fn from_integer(val: Scalar, fabric: SharedFabric<N, S>) -> Self {
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuit-types/src/fixed_point.rs:406]: It ended up not being needed, so I deleted the code.

File: circuit-macros/src/circuit_type/mpc_types.rs (hunk 3 above)
Comment (joeykraut) [reply circuit-macros/src/circuit_type/mpc_types.rs:140]: We could eventually make this part of the trait that the mpc types implement, but that's more complexity than is needed for now, I'll add a todo

File: circuit-types/src/traits.rs [hunk 5]
```diff
@@ -717,96 +688,104 @@ impl<const N: usize, T: CircuitCommitmentType> CircuitCommitmentType for [T; N]
...
 
-impl<N: MpcNetwork + Send, S: SharedValueSource<Scalar>> MpcType<N, S>
-    for AuthenticatedScalar<N, S>
-{
+impl MpcType for AuthenticatedScalarResult {
     type NativeType = Scalar;
 
-    fn from_authenticated_scalars<I: Iterator<Item = AuthenticatedScalar<N, S>>>(
+    fn fabric(&self) -> &MpcFabric {
+        self.fabric()
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuit-types/src/traits.rs:726]: The `self.fabric` resolves to a method named `fabric` on `AuthenticatedScalarResult`, not the trait

File: circuit-types/src/traits.rs [hunk 6]
```diff
@@ -193,21 +192,18 @@ pub trait CircuitCommitmentType: Clone {
...
-        fabric: SharedFabric<N, S>,
+        sender: PartyId,
+        fabric: MpcFabric,
     ) -> Result<Self::AllocatedType, MpcError> {
         let self_scalars = self.clone().to_scalars_with_linking();
-        let values = fabric
-            .borrow_fabric()
-            .batch_allocate_private_scalars(owning_party, &self_scalars)
-            .map_err(|err| MpcError::SetupError(err.to_string()))?;
+        let values = fabric.batch_share_scalar(self_scalars, sender);
```
[the remark this answers is not in the document]
Comment (joeykraut) [reply circuit-types/src/traits.rs:206]: correct

**Pull request merged**